### PR TITLE
(PUP-2280) Mark runs as failed when refresh events fail

### PR DESCRIPTION
**Pull request by Magisus**
Previously, if a 'refreshonly' exec failed, this failure was not
propagated into the report because refresh failures are tracked via a
different resource status, `failed_to_restart` as opposed to just
`failed`.

This commit updates the failure check logic in the transaction to
include checks of `failed_to_restart`, as well as a
similar update to the way the report status is calculated. A run is now
marked as failed if any of its resources are marked `failed_to_restart`.

### CODE REVIEW

File: lib/puppet/transaction.rb [hunk 1]
```diff
@@ -229,7 +231,7 @@ def prefetch_if_necessary(resource)
   def apply(resource, ancestor = nil)
     status = resource_harness.evaluate(resource)
     add_resource_status(status)
-    event_manager.queue_events(ancestor || resource, status.events) unless status.failed?
+    event_manager.queue_events(ancestor || resource, status.events) unless status.failed? || status.failed_to_restart?
```
Comment (MosesMendoza) [inline lib/puppet/transaction.rb]: this isn't necessary to the logic and its not your PR that makes this code hard to follow to be sure, but just for clarifying the flow could you add parens around the || conditional after unless? Or even just split it into multiple lines ie
```ruby
if !(status.failed? || status.failed_to_restart?)
  event_manager.queue_events(ancestor || resource, status.events)
end
```

Also the inline `||` in the param is a little too clever to follow IMO too...
what if we did something like:
```
ancestor ||= resource
event_manager.queue_events(ancestor, status.events)
...
```
Defer to your thoughts on it.

File: lib/puppet/transaction/report.rb [hunk 2]
```diff
@@ -176,7 +176,9 @@ def add_resource_status(status)
 
   # @api private
   def compute_status(resource_metrics, change_metric)
-    if resources_failed_to_generate || (resource_metrics["failed"] || 0) > 0
+    if resources_failed_to_generate ||
+       (resource_metrics["failed"] || 0) > 0 ||
+       (resource_metrics["failed_to_restart"] || 0) > 0
```
Comment (MosesMendoza) [inline lib/puppet/transaction/report.rb]: Can you add test coverage for this change?

File: lib/puppet/transaction.rb (hunk 1 above)
Comment (Magisus) [inline lib/puppet/transaction.rb]: Agreed, this is gross. I'll clean it up.

File: lib/puppet/transaction/report.rb (hunk 2 above)
Comment (Magisus) [inline lib/puppet/transaction/report.rb]: Done.

File: lib/puppet/transaction.rb (hunk 1 above)
Comment (Magisus) [inline lib/puppet/transaction.rb]: Udpated.

**Comment by MosesMendoza** [comment]
In the transaction_spec (ln 53) can you update the test description to include `failed_to_restart` ? 

Also probably good to add coverage for the change to `any_failed?`

**Comment by Magisus** [comment]
Do you have a suggestion for how best to trigger the `failed_to_restart` case for `any_failed`?

**Comment by MosesMendoza** [comment]
sure - evaluate the transaction, then set the resource status in the report to failed_to_restart. we should expect transaction.any_failed? to true (whereas prior to your PR it would be nil)

**Comment by Magisus** [comment]
Updated to add that tests. Confirmed it fails on master and passes with this PR.

**Comment by Iristyle** [comment]
👍
